includes/sanitizers/class-amp-form-sanitizer.php

```diff
@@ -67,23 +67,51 @@ public function sanitize() {
 				$action_url = esc_url_raw( '//' . $_SERVER['HTTP_HOST'] . wp_unslash( $_SERVER['REQUEST_URI'] ) );
 			} else {
 				$action_url = $node->getAttribute( 'action' );
-
-				// Handle relative URLs.
-				if ( ! preg_match( '#^(https?:)?//#', $action_url ) ) {
-					$schemeless_host = '//' . $_SERVER['HTTP_HOST'];
-					if ( '?' === $action_url[0] || '#' === $action_url[0] ) {
-						// For actions consisting of only a query or URL fragment, include the schemeless-host and the REQUEST URI of the current page.
-						$action_url = $schemeless_host . wp_unslash( $_SERVER['REQUEST_URI'] ) . $action_url;
-					} elseif ( '.' === $action_url[0] ) {
-						// For actions consisting of relative paths (e.g. '../'), prepend the schemeless-host and a trailing-slashed REQUEST URI.
-						$action_url = $schemeless_host . trailingslashit( wp_unslash( $_SERVER['REQUEST_URI'] ) ) . $action_url;
-					} else {
-						// Otherwise, when the action URL includes an absolute path, just append it to the schemeless-host.
-						$action_url = $schemeless_host . $action_url;
+				$parsed_url = wp_parse_url( $action_url );
+
+				// Ignore a malformed URL - it will be later sanitized.
+				if ( false !== $parsed_url ) {
+					// Handle relative URLs.
+					if ( ! isset( $parsed_url['scheme'] ) || 'https' !== $parsed_url['scheme'] ) {
+						$parsed_url['scheme'] = '//';
+
+						if ( ! isset( $parsed_url['host'] ) ) {
+							$parsed_url['host'] = $_SERVER['HTTP_HOST'];
+						}
+
+						if ( ! isset( $parsed_url['path'] ) ) {
+							// If there is action URL path, use the one from the request.
+							$parsed_url['path'] = trailingslashit( wp_unslash( $_SERVER['REQUEST_URI'] ) );
+						} elseif ( '' !== $parsed_url['path'] && '/' !== $parsed_url['path'][0] ) {
+							// If the path is relative, append it to the current request path.
+							$parsed_url['path'] = trailingslashit( wp_unslash( $_SERVER['REQUEST_URI'] ) ) . trailingslashit( $parsed_url['path'] );
+						}
```

Comment:
If no path is supplied and yet the host is present, then the path should assume to be empty. The current `REQUEST_URI` should not be supplied. I added a failing test case to show the proper behavior: f27f5a264251f4183ebfc35e94a79197d093aa07.

Reply:
Fixed in https://github.com/ampproject/amp-wp/pull/4250/commits/16b7fa8a7ee9b5ea7ebc47111f4495d8f6a54537.